toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -0,0 +1,584 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/semantics/semantics_ir_formatter.h"
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringSet.h"
+#include "llvm/Support/SaveAndRestore.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+#include "toolchain/parser/parse_tree.h"
+
+namespace Carbon {
+
+namespace {
+// Assigns names to nodes, blocks, and scopes in the Semantics IR.
+class NodeNamer {
+ public:
+  enum class ScopeIndex : int {
+    None = -1,
+    Package = 0,
+  };
+
+  NodeNamer(const TokenizedBuffer& tokenized_buffer,
+            const ParseTree& parse_tree, const SemanticsIR& semantics_ir)
+      : tokenized_buffer_(tokenized_buffer),
+        parse_tree_(parse_tree),
+        semantics_ir_(semantics_ir) {
+    nodes.resize(semantics_ir.nodes_size());
+    labels.resize(semantics_ir.node_blocks_size());
+    scopes.resize(1 + semantics_ir.functions_size());
+
+    // Build the package scope.
+    GetScopeInfo(ScopeIndex::Package).name = "package";
+    CollectNamesInBlock(ScopeIndex::Package, semantics_ir.top_node_block_id());
+
+    // Build each function scope.
+    for (int i = 0; i != semantics_ir.functions_size(); ++i) {
+      auto fn_id = SemanticsFunctionId(i);
+      auto fn_scope = GetScopeFor(fn_id);
+      const auto& fn = semantics_ir.GetFunction(fn_id);
+      GetScopeInfo(fn_scope).name =
+          "@" +
+          globals.AllocateName(*this,
+                               /*TODO*/ ParseTree::Node::Invalid,
+                               fn.name_id.is_valid()
+                                   ? semantics_ir.GetString(fn.name_id).str()
+                                   : "");
+      CollectNamesInBlock(fn_scope, fn.param_refs_id);
+      for (auto block_id : fn.body_block_ids) {
+        AddBlockLabel(fn_scope, block_id,
+                      block_id == fn.body_block_ids.front() ? "entry" : "");
+        CollectNamesInBlock(fn_scope, block_id);
+      }
+    }
+  }
+
+  // Returns the scope index corresponding to a function.
+  auto GetScopeFor(SemanticsFunctionId fn_id) -> ScopeIndex {
+    return ScopeIndex(fn_id.index + 1);
+  }
+
+  // Returns the IR name to use for a function.
+  auto GetNameFor(SemanticsFunctionId fn_id) -> std::string {
+    if (!fn_id.is_valid()) {
+      return "invalid";
+    }
+    return GetScopeInfo(GetScopeFor(fn_id)).name;
+  }
+
+  // Returns the IR name to use for a node, when referenced from a given scope.
+  auto GetNameFor(ScopeIndex scope_idx, SemanticsNodeId node_id)
+      -> std::string {
+    if (!node_id.is_valid()) {
+      return "invalid";
+    }
+
+    // Check for a builtin.
+    if (node_id.index < SemanticsBuiltinKind::ValidCount) {
+      return SemanticsBuiltinKind::FromInt(node_id.index).label().str();
+    }
+
+    auto& [node_scope, node_name] = nodes[node_id.index];
+    if (node_name.empty()) {
+      // This should not happen in valid IR.
+      return "<noderef " + llvm::itostr(node_id.index) + ">";
```

Comment:
What do you think about here and elsewhere something like:

```suggestion
      // This should not happen in valid IR.
      return "<invalid noderef " + llvm::itostr(node_id.index) + ">";
```

Reply:
Done, went with "unexpected" rather than "invalid" because the node reference itself is valid (references a valid node), but not one that we expected to be referenced by ID.